tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,519 @@
+    /// }
+    /// ```
+    pub fn notify(&self) {
+        // Load the current state
+        let mut curr = self.state.load(SeqCst);
+
+        // If the state is `EMPTY`, transition to `NOTIFIED` and return.
+        while let EMPTY | NOTIFIED = curr {
+            // The compare-exchange from `NOTIFIED` -> `NOTIFIED` is intended. A
+            // happens-before synchronization must happen between this atomic
+            // operation and a task calling `notified().await`.
+            let res = self.state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+            match res {
+                // No waiters, no further work to do
+                Ok(_) => return,
+                Err(actual) => {
+                    curr = actual;
+                }
+            }
+        }
+
+        // There are waiters, the lock must be acquired to notify.
+        let mut waiters = self.waiters.lock().unwrap();
+
+        // The state must be reloaded while the lock is held. The state may only
+        // transition out of WAITING while the lock is held.
+        curr = self.state.load(SeqCst);
+
+        if let Some(waker) = notify_locked(&mut waiters, &self.state, curr) {
+            drop(waiters);
+            waker.wake();
+        }
+    }
+}
+
+impl Default for Notify {
+    fn default() -> Notify {
+        Notify::new()
+    }
+}
+
+fn notify_locked(waiters: &mut LinkedList<Waiter>, state: &AtomicU8, curr: u8) -> Option<Waker> {
+    loop {
+        match curr {
+            EMPTY | NOTIFIED => {
+                let res = state.compare_exchange(curr, NOTIFIED, SeqCst, SeqCst);
+
+                match res {
+                    Ok(_) => return None,
+                    Err(actual) => {
+                        assert!(actual == EMPTY || actual == NOTIFIED);
+                        state.store(NOTIFIED, SeqCst);
+                        return None;
+                    }
+                }
+            }
+            WAITING => {
+                // At this point, it is guaranteed that the state will not
+                // concurrently change as holding the lock is required to
+                // transition **out** of `WAITING`.
+                //
+                // Get a pending waiter
+                let mut waiter = waiters.pop_back().unwrap();
+
+                assert!(!waiter.notified);
+
+                waiter.notified = true;
+                let waker = waiter.waker.take();
+
+                if waiters.is_empty() {
+                    // As this the **final** waiter in the list, the state
+                    // must be transitioned to `EMPTY`. As transitioning
+                    // **from** `WAITING` requires the lock to be held, a
+                    // `store` is sufficient.
+                    state.store(EMPTY, SeqCst);
+                }
+
+                return waker;
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+// ===== impl Notified =====
+
+impl Notified<'_> {
+    fn project(
```

Comment:
nit: might be worth commenting that we have to implement a manual pin projection for this type because `pin_project_lite` doesn't work with types implementing `Drop`?